File: 2015/cpp/day21/part2.cpp
#include <iostream>
#include <stdlib.h>

using namespace std;

int main () {
    int health = 100;
    int bossHealth = 104;
    int bossDamage = 8;
    int bossArmor = 1;

    int weapons[][5] = {
        {8, 4},
        {10, 5},
        {25, 6},
        {40, 7},
        {74, 8}
    };
    int armors [][6] = {
        {0, 0},
        {13, 1},
        {31, 2},
        {53, 3},
        {75, 4},
        {102, 5}
    };
    int rings[][8] = {
        {0, 0, 0},
        {0, 0, 0},
        {25, 1, 0},
        {50, 2, 0},
        {100, 3, 0},
        {20, 0, 1},
        {40, 0, 2},
        {80, 0, 3}
    };

    int record = 0;

    for (int weapon = 0; weapon < 5; weapon++) {
        for (int armor = 0; armor < 6; armor++) {
            for (int ring1 = 0; ring1 < 8; ring1++) {
                for (int ring2 = 0; ring2 < 8; ring2++) {
                    if (ring1 == ring2) continue;

                    int gold = weapons[weapon][0] + armors[armor][0] +
                                rings[ring1][0] + rings[ring2][0];

                    if (gold < record) continue;

                    int damage = weapons[weapon][1] + rings[ring1][1] + rings[ring2][1];
                    int defense = armors[armor][1] + rings[ring1][2] + rings[ring2][2];

                    int hitsToKillBoss = (bossHealth - 1) / max(damage - bossArmor, 1);
                    int hitsToKillPlayer = (health - 1) / max(bossDamage - defense, 1);

                    if (hitsToKillPlayer < hitsToKillBoss)
                        record = max(record, gold);
                }
            }
        }
    }

    cout << record << endl;
}

File: 2015/ruby/day21/part2.rb
#!/usr/bin/ruby

boss = {
    :health => 104,
    :damage => 8,
    :armor => 1
}

weapons = [
    {:cost => 8,  :damage => 4},
    {:cost => 10, :damage => 5},
    {:cost => 25, :damage => 6},
    {:cost => 40, :damage => 7},
    {:cost => 74, :damage => 8}
]

armor = [
    {:cost => 0,   :armor => 0},
    {:cost => 13,  :armor => 1},
    {:cost => 31,  :armor => 2},
    {:cost => 53,  :armor => 3},
    {:cost => 75,  :armor => 4},
    {:cost => 102, :armor => 5}
]

rings = [
    {:cost => 0,   :damage => 0, :armor => 0},
    {:cost => 0,   :damage => 0, :armor => 0},
    {:cost => 25,  :damage => 1, :armor => 0},
    {:cost => 50,  :damage => 2, :armor => 0},
    {:cost => 100, :damage => 3, :armor => 0},
    {:cost => 20,  :damage => 0, :armor => 1},
    {:cost => 40,  :damage => 0, :armor => 2},
    {:cost => 80,  :damage => 0, :armor => 3}
]

most_spent = 0

weapons.each do |weapon|
    armor.each do |armor|
        rings.each do |ring1|
            rings.each do |ring2|
                if ring1[:cost] == ring2[:cost] and ring1[:cost] != 0
                    next
                end

                player = {
                    :health => 100,
                    :damage => weapon[:damage] + ring1[:damage] + ring2[:damage],
                    :armor => armor[:armor] + ring1[:armor] + ring2[:armor]
                }

                hits_to_kill_player = (player[:health] - 1) / [boss[:damage] - player[:armor], 1].max
                hits_to_kill_boss = (boss[:health] - 1) / [player[:damage] - boss[:armor], 1].max

                if hits_to_kill_player < hits_to_kill_boss
                    cost = weapon[:cost] + armor[:cost] + ring1[:cost] + ring2[:cost]
                    most_spent = cost > most_spent ? cost : most_spent
                end
            end
        end
    end
end

puts most_spent
